src/Symfony/Component/Notifier/Notifier.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Notifier\Channel\ChannelInterface;
+use Symfony\Component\Notifier\Channel\ChannelPolicy;
+use Symfony\Component\Notifier\Channel\ChannelPolicyInterface;
+use Symfony\Component\Notifier\Exception\LogicException;
+use Symfony\Component\Notifier\Notification\Notification;
+use Symfony\Component\Notifier\Receiver\AdminReceiver;
+use Symfony\Component\Notifier\Receiver\Receiver;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @experimental in 5.0
+ */
+final class Notifier implements NotifierInterface
+{
+    private static $adminReceivers = [];
+
+    private $channels;
+    private $policy;
+    private $bus;
+
+    /**
+     * @param ChannelInterface[]|ContainerInterface $channels
+     */
+    public function __construct($channels, ChannelPolicyInterface $policy = null, MessageBusInterface $bus = null)
+    {
+        $this->channels = $channels;
+        $this->policy = $policy;
+        $this->bus = $bus;
+    }
+
+    public function send(Notification $notification, Receiver ...$receivers): void
+    {
+        foreach ($receivers as $receiver) {
+            foreach ($this->getChannels($notification, $receiver) as $channel => $transport) {
+                $channel->notify($notification, $receiver, $transport, $this->bus);
+            }
+        }
+    }
+
+    public static function addAdminReceiver(AdminReceiver $receiver): void
+    {
+        self::$adminReceivers[] = $receiver;
+    }
+
+    /**
+     * @return AdminReceiver[]
+     */
+    public static function getAdminReceivers(): array
+    {
+        return self::$adminReceivers;
+    }
+
+    private function getChannels(Notification $notification, Receiver $receiver): iterable
+    {
+        $channels = $notification->getChannels($receiver);
+        if (!$channels) {
+            if (null === $this->policy) {
+                throw new LogicException(sprintf('Unable to determine which channels to use to send a notification; you should either override the "getChannels()" method on "%s" or configure a "%s".', \get_class($notification), ChannelPolicy::class));
+            }
+            $channels = $this->policy->getChannels($notification->getImportance());
+        }
+
+        if (!$channels) {
+            throw new LogicException(sprintf('Unable to determine which channels to use to send a notification; you should either override the "getChannels()" method on "%s" or configure a "%s".', \get_class($notification), ChannelPolicy::class));
+        }
+
+        foreach ($channels as $channel) {
```

Comment:
I would rename `channel` as `channelName` and so `c` as `channel` ... in the whole component. I know this is a lot a work, but this would avoid confusion between a channel name, and a channel implementation

Reply:
changed in this class as we manipulate both